### PR TITLE
Improve mypy for azure-storage-blob

### PR DESCRIPTION
**Pull request by rakshith91**
Resolves https://github.com/Azure/azure-sdk-for-python/issues/10859

### CODE REVIEW

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client_async.py [hunk 1]
```diff
@@ -66,25 +68,25 @@ async def close(self):
...
         elif isinstance(credential, SharedKeyCredentialPolicy):
-            self._credential_policy = credential
+            self._credential_policy = cast(T, credential)
         elif credential is not None:
             raise TypeError("Unsupported credential: {}".format(credential))
         config = kwargs.get('_configuration') or create_configuration(**kwargs)
         if kwargs.get('_pipeline'):
             return config, kwargs['_pipeline']
-        config.transport = kwargs.get('transport')  # type: ignore
+        _transport = kwargs.get("transport")
```
Comment (lmazuel) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client_async.py:81]: Breaking change, you change the behavior of `config.transport` which is `None` now

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client_async.py [hunk 2]
```diff
@@ -40,6 +41,7 @@
     from azure.core.pipeline import Pipeline
     from azure.core.pipeline.transport import HttpRequest
     from azure.core.configuration import Configuration
+T = Optional[Union[BearerTokenCredentialPolicy, SharedKeyCredentialPolicy]]
```
Comment (lmazuel) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client_async.py:44]: Rename it to something more explicit, it gives the impression that it's a generic, but it's not. Example `OptionalStorageCredentials`

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py [hunk 3]
```diff
@@ -217,14 +158,13 @@ def from_blob_url(cls, blob_url, credential=None, snapshot=None, **kwargs):
             raise ValueError("Invalid URL. Provide a blob_url with a valid blob and container name.")
 
         path_snapshot, _ = parse_query(parsed_url.query)
-        if snapshot:
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py:220]: >snapshot [](start = 11, length = 8)

if we delete this, then will None.snapshot throw AttributeError or other error type?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py [hunk 4]
```diff
@@ -1785,19 +1273,10 @@ def stage_block_from_url(
...
-    def _get_block_list_result(self, blocks):
-        # type: (BlockList) -> Tuple[List[BlobBlock], List[BlobBlock]]
-        committed = [] # type: List
-        uncommitted = [] # type: List
-        if blocks.committed_blocks:
-            committed = [BlobBlock._from_generated(b) for b in blocks.committed_blocks]  # pylint: disable=protected-access
-        if blocks.uncommitted_blocks:
-            uncommitted = [BlobBlock._from_generated(b) for b in blocks.uncommitted_blocks]  # pylint: disable=protected-access
-        return committed, uncommitted
+        return cast(Dict, ret_val)
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py:1279]: >return cast(Dict, ret_val) [](start = 8, length = 26)

just curious, will mypy be angry if you we put this return inside try block?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py [hunk 5]
```diff
@@ -1967,9 +1383,11 @@ def commit_block_list( # type: ignore
             metadata=metadata,
             **kwargs)
         try:
-            return self._client.block_blob.commit_block_list(**options) # type: ignore
+            ret_val = self._client.block_blob.commit_block_list(**options)
         except StorageErrorException as error:
+            ret_val = None
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py:1388]: > ret_val = None [](start = 11, length = 15)

It seems this line could be deleted?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py [hunk 6]
```diff
@@ -2237,33 +1606,10 @@ def set_sequence_number(self, sequence_number_action, sequence_number=None, **kw
...
-        options = {
-            'blob_content_length': size,
-            'timeout': kwargs.pop('timeout', None),
-            'lease_access_conditions': access_conditions,
-            'modified_access_conditions': mod_conditions,
-            'cpk_info': cpk_info,
-            'cls': return_response_headers}
-        options.update(kwargs)
-        return options
+        return cast(Dict, None)
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_blob_client.py:1612]: >   return cast(Dict, None) [](start = 5, length = 26)

what will happen if we delete this line?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_lease.py [hunk 7]
```diff
@@ -121,9 +105,9 @@ def acquire(self, lease_duration=-1, **kwargs):
                 **kwargs)
         except StorageErrorException as error:
             process_storage_error(error)
-        self.id = response.get('lease_id')  # type: str
-        self.last_modified = response.get('last_modified')   # type: datetime
-        self.etag = kwargs.get('etag')  # type: str
+        self.id = response.get('lease_id')
+        self.last_modified = response.get('last_modified')
+        self.etag = kwargs.get('etag')
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_lease.py:110]: would you like to make it `self.etag = response.get('etag')`, I feel it's a 'typo' in the existing code

File: sdk/storage/azure-storage-blob/azure/storage/blob/_lease_client_base.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import uuid
+
+from typing import (  # pylint: disable=unused-import
+    Union, Optional, Any, IO, Iterable, AnyStr, Dict, List, Tuple, cast,
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_lease_client_base.py:10]: >cast [](start = 67, length = 4)

some of them are not used?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_lease_client_base.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+if TYPE_CHECKING:
+    from datetime import datetime
+    from ._generated.operations import BlobOperations, ContainerOperations
+    BlobClient = TypeVar("BlobClient")
+    ContainerClient = TypeVar("ContainerClient")
+
+
+class BlobLeaseClientBase(object):
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_lease_client_base.py:21]: Do we extract this class out just for consistency? what will happen if we leave this class inside _lease.py?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client.py [hunk 10]
```diff
@@ -65,7 +75,7 @@
     "file": {"primary": "FileEndpoint", "secondary": "FileSecondaryEndpoint"},
     "dfs": {"primary": "BlobEndpoint", "secondary": "BlobEndpoint"},
 }
-
+T = Optional[Union[BearerTokenCredentialPolicy, SharedKeyCredentialPolicy]]
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client.py:78]: can we rename T to another name

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client.py [hunk 11]
```diff
@@ -224,11 +234,11 @@ def _create_pipeline(self, credential, **kwargs):
...
+        _transport = kwargs.get("transport")
         kwargs.setdefault("connection_timeout", CONNECTION_TIMEOUT)
         kwargs.setdefault("read_timeout", READ_TIMEOUT)
-        if not config.transport:
-            config.transport = RequestsTransport(**kwargs)
+        if not _transport:
+            _transport = RequestsTransport(**kwargs)
         policies = [
             QueueMessagePolicy(),
             config.headers_policy,
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client.py:244]: >config [](start = 12, length = 6)

so now the config doesn't have attribute `transport`?

**Comment by xiafu-msft** [comment]
sorry It seems it was automatically closed when I merged stg73 to master

**Pull request closed without merging**